### PR TITLE
[MXNET-382] Shape and Size Operator

### PR DESCRIPTION
**Pull request by anirudhacharya**
## Description ##
Shape and Size Operator. Pre-requisite for this issue - https://github.com/apache/incubator-mxnet/issues/10789

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [x] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [x] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [x] Shape Operator
- [x] Size Operator

@haojin2 @eric-haibin-lin @zheng-da

### CODE REVIEW

File: src/operator/tensor/elemwise_unary_op_basic.cc [hunk 1]
```diff
@@ -399,6 +399,37 @@ NNVM_REGISTER_OP(reshape_like)
...
+.set_attr<nnvm::FInferType>("FInferType",
+	[](const nnvm::NodeAttrs& attrs,
+	   std::vector<int>* in_attrs,
+	   std::vector<int>* out_attrs) {
+		CHECK_EQ(in_attrs->size(), 1U);
+	    CHECK_EQ(out_attrs->size(), 1U);
+	    TYPE_ASSIGN_CHECK(*out_attrs, 0, 0U);
+	    return out_attrs->at(0) != -1;
+		})
+.set_attr<nnvm::FGradient>("FGradient", MakeZeroGradNodes)
```
Comment (reminisce) [inline src/operator/tensor/elemwise_unary_op_basic.cc]: Shouldn't there be no appropriate gradient definition for this?

File: src/operator/tensor/elemwise_unary_op.h [hunk 2]
```diff
@@ -388,6 +388,26 @@ void CastCompute(const nnvm::NodeAttrs& attrs,
   });
 }
 
+template<typename xpu>
+void ShapeCompute(const nnvm::NodeAttrs& attrs,
+                 const OpContext& ctx,
```
Comment (reminisce) [inline src/operator/tensor/elemwise_unary_op.h]: alignment.

File: src/operator/tensor/elemwise_unary_op.h [hunk 3]
```diff
@@ -388,6 +388,26 @@ void CastCompute(const nnvm::NodeAttrs& attrs,
...
+  CHECK_EQ(inputs.size(), 1U);
+  CHECK_EQ(outputs.size(), 1U);
+  CHECK_EQ(req.size(), 1U);
+  const TBlob& in_data = inputs[0];
+  const TBlob& out_data = outputs[0];
+  mshadow::Stream<xpu> *s = ctx.get_stream<xpu>();
+  TShape in_shape = in_data.shape_;
+  MSHADOW_TYPE_SWITCH(out_data.type_flag_, DType, {
+    mxnet_op::Kernel<mshadow_op::identity_with_cast, xpu>::Launch(
+	  s, in_data.ndim(), out_data.dptr<DType>(), in_shape.data());
```
Comment (reminisce) [inline src/operator/tensor/elemwise_unary_op.h]: Why `DType` for `out_data`? Isn't that `int64` as in the description?

File: src/operator/tensor/elemwise_unary_op.h [hunk 4]
```diff
@@ -388,6 +388,26 @@ void CastCompute(const nnvm::NodeAttrs& attrs,
...
+                 const std::vector<TBlob>& inputs,
+                 const std::vector<OpReqType>& req,
+                 const std::vector<TBlob>& outputs) {
+  CHECK_EQ(inputs.size(), 1U);
+  CHECK_EQ(outputs.size(), 1U);
+  CHECK_EQ(req.size(), 1U);
+  const TBlob& in_data = inputs[0];
+  const TBlob& out_data = outputs[0];
+  mshadow::Stream<xpu> *s = ctx.get_stream<xpu>();
+  TShape in_shape = in_data.shape_;
```
Comment (reminisce) [inline src/operator/tensor/elemwise_unary_op.h]: const TShape&.

File: src/operator/tensor/elemwise_unary_op_basic.cc [hunk 5]
```diff
@@ -399,6 +399,37 @@ NNVM_REGISTER_OP(reshape_like)
 .add_argument("lhs", "NDArray-or-Symbol", "First input.")
 .add_argument("rhs", "NDArray-or-Symbol", "Second input.");
 
+NNVM_REGISTER_OP(shape)
+.describe("Returns a 1D int64 array containing the shape of data.")
+.set_num_inputs(1)
+.set_num_outputs(1)
+.set_attr<nnvm::FInplaceIdentity>("FInplaceIdentity",
```
Comment (reminisce) [inline src/operator/tensor/elemwise_unary_op_basic.cc]: Why add this?

File: src/operator/tensor/elemwise_unary_op_basic.cc [hunk 6]
```diff
@@ -399,6 +399,37 @@ NNVM_REGISTER_OP(reshape_like)
...
+.set_attr<nnvm::FInferShape>("FInferShape",
+	[](const nnvm::NodeAttrs& attrs,
+	   std::vector<TShape> *in_attrs,
+	   std::vector<TShape> *out_attrs) {
+	     CHECK_EQ(in_attrs->size(), 1U);
+	     CHECK_EQ(out_attrs->size(), 1U);
+	     TShape target_shape(1);
+	     target_shape[0] = in_attrs->at(0).ndim();
+	     SHAPE_ASSIGN_CHECK(*out_attrs, 0, target_shape);
+	     return out_attrs->at(0).ndim() != 0U && out_attrs->at(0).Size() != 0U;
```
Comment (reminisce) [inline src/operator/tensor/elemwise_unary_op_basic.cc]: use `shape_is_none(out_attrs->at(0)`

File: src/operator/tensor/elemwise_unary_op_basic.cc [hunk 7]
```diff
@@ -399,6 +399,37 @@ NNVM_REGISTER_OP(reshape_like)
...
+	    })
+.set_attr<nnvm::FInferType>("FInferType",
+	[](const nnvm::NodeAttrs& attrs,
+	   std::vector<int>* in_attrs,
+	   std::vector<int>* out_attrs) {
+		CHECK_EQ(in_attrs->size(), 1U);
+	    CHECK_EQ(out_attrs->size(), 1U);
+	    TYPE_ASSIGN_CHECK(*out_attrs, 0, 0U);
```
Comment (reminisce) [inline src/operator/tensor/elemwise_unary_op_basic.cc]: Use type enum variable name.

File: src/operator/tensor/elemwise_unary_op_basic.cu [hunk 8]
```diff
@@ -77,6 +77,9 @@ NNVM_REGISTER_OP(_identity_with_attr_like_rhs)
 NNVM_REGISTER_OP(reshape_like)
 .set_attr<FCompute>("FCompute<gpu>", UnaryOp::IdentityCompute<gpu>);
 
+NNVM_REGISTER_OP(shape)
```
Comment (reminisce) [inline src/operator/tensor/elemwise_unary_op_basic.cu]: Where does the name come from? This looks confusing and conflicts with the property name `shape` of `NDArray` in Python. Need to ensure the documentation can be rendered correctly.

Comment (anirudhacharya) [reply src/operator/tensor/elemwise_unary_op_basic.cu]: what do you suggest the name of the operator should be? This is more or less the name used in a couple of other frameworks too.

File: src/operator/tensor/elemwise_unary_op_basic.cc (hunk 5 above)
Comment (anirudhacharya) [reply src/operator/tensor/elemwise_unary_op_basic.cc]: will remove it.

File: src/operator/tensor/elemwise_unary_op_basic.cc (hunk 6 above)
Comment (anirudhacharya) [reply src/operator/tensor/elemwise_unary_op_basic.cc]: you suggest that I define a shape_is_none function and use it here?

File: src/operator/tensor/elemwise_unary_op_basic.cc (hunk 7 above)
Comment (anirudhacharya) [reply src/operator/tensor/elemwise_unary_op_basic.cc]: okay, i will use mshadow::kInt64

**Comment by zheng-da** [comment]
the shape operator doesn't have backward. my biggest concern is that any computation that uses this operator can't perform a backward computation.

File: src/operator/tensor/elemwise_unary_op.h [hunk 9]
```diff
@@ -388,6 +388,26 @@ void CastCompute(const nnvm::NodeAttrs& attrs,
...
+  MSHADOW_TYPE_SWITCH(out_data.type_flag_, DType, {
+    mxnet_op::Kernel<mshadow_op::identity_with_cast, xpu>::Launch(
+	  s, in_data.ndim(), out_data.dptr<DType>(), in_shape.data());
+  });
+}
+
+
```
Comment (haojin2) [inline src/operator/tensor/elemwise_unary_op.h]: Please get rid of one blank line here, c++ use only 1 blank line between functions

File: src/operator/tensor/elemwise_unary_op_basic.cc (hunk 1 above)
Comment (anirudhacharya) [reply src/operator/tensor/elemwise_unary_op_basic.cc]: no, shape and size operator will not have gradients right?

File: src/operator/tensor/elemwise_unary_op_basic.cc [hunk 10]
```diff
@@ -399,6 +399,37 @@ NNVM_REGISTER_OP(reshape_like)
 .add_argument("lhs", "NDArray-or-Symbol", "First input.")
 .add_argument("rhs", "NDArray-or-Symbol", "Second input.");
 
+NNVM_REGISTER_OP(shape)
+.describe("Returns a 1D int64 array containing the shape of data.")
```
Comment (haojin2) [inline src/operator/tensor/elemwise_unary_op_basic.cc]: Would it be clearer if you add an example here?

File: src/operator/tensor/elemwise_unary_op_basic.cc (hunk 1 above)
Comment (reminisce) [reply src/operator/tensor/elemwise_unary_op_basic.cc]: Right, I'm asking the same thing. Why did you define FGradient attribute for the op?

File: src/operator/tensor/elemwise_unary_op_basic.cc (hunk 6 above)
Comment (reminisce) [reply src/operator/tensor/elemwise_unary_op_basic.cc]: It's already available.

File: src/operator/tensor/elemwise_unary_op_basic.cu (hunk 8 above)
Comment (reminisce) [reply src/operator/tensor/elemwise_unary_op_basic.cu]: Please make sure the doc page can be rendered correctly at least.

**Comment by chinakook** [comment]
That's great. Keras and Tensorflow both have this op.

**Comment by piiswrong** [comment]
The name shape_op and size_op looks adhoc

**Comment by anirudhacharya** [comment]
@piiswrong shape and size are already existing ndarray operations. I changed the names to prevent confusion. Some of the other names that come to mind are - shape_operator, tensor_shape, array_shape.

**Comment by reminisce** [comment]
How about `shape_nd` and `size_nd`, which mimics the naming of `gather_nd` and `scatter_nd`?

**Comment by reminisce** [comment]
Please also add tests here:
https://github.com/apache/incubator-mxnet/blob/master/tests/python/unittest/test_ndarray.py#L913
https://github.com/apache/incubator-mxnet/blob/master/tests/python/unittest/test_symbol.py#L171

File: src/operator/tensor/elemwise_unary_op_basic.cc [hunk 11]
```diff
@@ -399,6 +399,71 @@ NNVM_REGISTER_OP(reshape_like)
...
+      std::vector<TShape> *in_attrs,
+      std::vector<TShape> *out_attrs) {
+       CHECK_EQ(in_attrs->size(), 1U);
+       CHECK_EQ(out_attrs->size(), 1U);
+       TShape target_shape(1);
+       target_shape[0] = in_attrs->at(0).ndim();
+       SHAPE_ASSIGN_CHECK(*out_attrs, 0, target_shape);
+       return !shape_is_none(out_attrs->at(0));
+     })
+.set_attr<nnvm::FInferType>("FInferType",
```
Comment (anirudh2290) [inline src/operator/tensor/elemwise_unary_op_basic.cc]: not registering FGradient ?

Comment (anirudhacharya) [reply src/operator/tensor/elemwise_unary_op_basic.cc]: Shape operator does not have a differential. Check the conversation here - https://github.com/apache/incubator-mxnet/issues/10789#issuecomment-388274133

File: src/operator/tensor/elemwise_unary_op.h [hunk 12]
```diff
@@ -388,6 +388,39 @@ void CastCompute(const nnvm::NodeAttrs& attrs,
...
+                 const OpContext& ctx,
+                 const std::vector<TBlob>& inputs,
+                 const std::vector<OpReqType>& req,
+                 const std::vector<TBlob>& outputs) {
+  CHECK_EQ(inputs.size(), 1U);
+  CHECK_EQ(outputs.size(), 1U);
+  CHECK_EQ(req.size(), 1U);
+  const TBlob& in_data = inputs[0];
+  const TBlob& out_data = outputs[0];
+  out_data.dptr<int64_t>()[0] = in_data.Size();
```
Comment (eric-haibin-lin) [inline src/operator/tensor/elemwise_unary_op.h]: out_data holds a pointer to gpu memory. you need to explicitly use kernel launch to set the value

Comment (anirudhacharya) [reply src/operator/tensor/elemwise_unary_op.h]: If I use kernel launch then I will need a databuffer pointing to in_data.Size(). How would I get that? because in_data.Size() is of type ``index_t`` which does not have ``data`` or ``dptr`` attribute.

Comment (haojin2) [reply src/operator/tensor/elemwise_unary_op.h]: You know the output size is only 1, so you can just use 1 for that.

Comment (anirudhacharya) [reply src/operator/tensor/elemwise_unary_op.h]:
```
mxnet_op::Kernel<mshadow_op::identity_with_cast, xpu>::Launch(s, 1U, out_data.dptr<int64_t>(), < what goes here? - in_data.Size()?? >);
```

**Comment by szha** [comment]
Ping for another round of reviews.

File: src/operator/mshadow_op.h [hunk 13]
```diff
@@ -96,6 +96,12 @@ struct identity_with_cast {
   }
 };
 
+struct size_kernel {
+  MSHADOW_XINLINE static void Map(int i, int64_t *out, unsigned int in) {
+    out[0] = int64_t(in);
```
Comment (reminisce) [inline src/operator/mshadow_op.h]: nit: `static_cast<int64_t>`

File: src/operator/tensor/elemwise_unary_op.h [hunk 14]
```diff
@@ -388,6 +388,43 @@ void CastCompute(const nnvm::NodeAttrs& attrs,
...
+  CHECK_EQ(inputs.size(), 1U);
+  CHECK_EQ(outputs.size(), 1U);
+  CHECK_EQ(req.size(), 1U);
+  const TBlob& in_data = inputs[0];
+  const TBlob& out_data = outputs[0];
+  mshadow::Stream<xpu> *s = ctx.get_stream<xpu>();
+  const TShape& in_shape = in_data.shape_;
+  MSHADOW_TYPE_SWITCH(out_data.type_flag_, DType, {
+    mxnet_op::Kernel<mshadow_op::identity_with_cast, xpu>::Launch(
+      s, in_data.ndim(), out_data.dptr<int64_t>(), in_shape.data());
```
Comment (reminisce) [inline src/operator/tensor/elemwise_unary_op.h]: `in_shape.data` is a pointer in cpu memory which cannot be directly accessed on gpu. You can use `Shape<ndim>` instead.

**Comment by piiswrong** [comment]
shape_nd still sounds weird as it's also available in symbol.

BTW I think these operators can be useful but they won't solve the issue #10789

**Comment by anirudhacharya** [comment]
@piiswrong Yes, it will not solve the issue. We will also need changes to define the gradient of a shape operator. Currently shape and size have no gradient defined. Also we will need changes in nnvm to propagate this information during back-propagation.

File: src/operator/tensor/elemwise_unary_op.h (hunk 14 above)
Comment (piiswrong) [reply src/operator/tensor/elemwise_unary_op.h]: how come this is not captured by CI?

**Comment by piiswrong** [comment]
how about shape_array and size_array

**Comment by anirudhacharya** [comment]
@piiswrong sure, but what about consistency? Isn't it a bit confusing for users if we have [gather_nd](http://mxnet.incubator.apache.org/api/python/symbol/symbol.html#mxnet.symbol.gather_nd) [scatter_nd](http://mxnet.incubator.apache.org/api/python/symbol/symbol.html#mxnet.symbol.scatter_nd) and then ``shape_array`` and ``size_array``. We should have some sort of naming policy so that the same naming conventions can also be used in the future.

File: src/operator/tensor/elemwise_unary_op.h (hunk 14 above)
Comment (anirudhacharya) [reply src/operator/tensor/elemwise_unary_op.h]: it did, the CI failed for GPU tests. I need to fix it.

**Comment by piiswrong** [comment]
gather_nd has a completely different context. It's in contrast to gather.

**Comment by anirudhacharya** [comment]
@reminisce ping for review.

File: tests/python/unittest/test_symbol.py [hunk 15]
```diff
@@ -190,8 +190,8 @@ def check_fluent_regular(func, kwargs, shape=(5, 17, 1), equal_nan=False):
                                      equal_nan=equal_nan)
 
     for func in ['flatten', 'norm', 'round', 'rint', 'fix', 'floor', 'ceil', 'trunc', 'zeros_like',
-                 'ones_like', 'abs', 'sign', 'sin', 'cos', 'degrees', 'radians',
-                 'exp', 'expm1',  'square', 'reciprocal', 'argmax_channel']:
+                 'ones_like', 'abs', 'sign', 'sin', 'cos', 'degrees', 'radians','exp', 'expm1',
```
Comment (haojin2) [inline tests/python/unittest/test_symbol.py]: missing a space before `'exp'`

File: tests/python/unittest/test_ndarray.py [hunk 16]
```diff
@@ -932,8 +932,8 @@ def check_fluent_regular(func, kwargs, shape=(5, 17, 1), equal_nan=False):
                 assert almost_equal(regular.asnumpy(), fluent.asnumpy(), equal_nan=equal_nan)
 
     for func in ['flatten', 'norm', 'round', 'rint', 'fix', 'floor', 'ceil', 'trunc', 'zeros_like',
-                 'ones_like', 'abs', 'sign', 'sin', 'cos', 'degrees', 'radians',
-                 'exp', 'expm1', 'square', 'reciprocal', 'argmax_channel']:
+                 'ones_like', 'abs', 'sign', 'sin', 'cos', 'degrees', 'radians','exp', 'expm1',
```
Comment (haojin2) [inline tests/python/unittest/test_ndarray.py]: missing a space before `'exp'`

File: src/operator/tensor/elemwise_unary_op_basic.cc [hunk 17]
```diff
@@ -398,6 +398,106 @@ NNVM_REGISTER_OP(reshape_like)
...
+      s, in_data.ndim(), out_data.dptr<DType>(), in_shape.data());
+  });
+}
+
+NNVM_REGISTER_OP(shape_array)
+.describe(R"code(Returns a 1D int64 array containing the shape of data.
+
+Example::
+
+  shape_nd([[1,2,3,4], [5,6,7,8]]) = [2,4]
```
Comment (reminisce) [inline src/operator/tensor/elemwise_unary_op_basic.cc]: shape_array?

File: src/operator/tensor/elemwise_unary_op_basic.cc [hunk 18]
```diff
@@ -398,6 +398,106 @@ NNVM_REGISTER_OP(reshape_like)
...
+                      const std::vector<TBlob>& outputs) {
+  using namespace mshadow;
+  using namespace mxnet_op;
+  CHECK_EQ(inputs.size(), 1U);
+  CHECK_EQ(outputs.size(), 1U);
+  CHECK_EQ(req.size(), 1U);
+  const TBlob& in_data = inputs[0];
+  const TBlob& out_data = outputs[0];
+  mshadow::Stream<cpu> *s = ctx.get_stream<cpu>();
+  mxnet_op::Kernel<mshadow_op::size_kernel, cpu>::Launch(
```
Comment (reminisce) [inline src/operator/tensor/elemwise_unary_op_basic.cc]: If it's just simply copying one element, no need to use kernel launch scheme.

File: src/operator/tensor/elemwise_unary_op_basic.cu [hunk 19]
```diff
@@ -77,6 +77,56 @@ NNVM_REGISTER_OP(_identity_with_attr_like_rhs)
...
+  const TBlob& out_data = outputs[0];
+  mshadow::Stream<gpu> *s = ctx.get_stream<gpu>();
+  const TShape& in_shape = in_data.shape_;
+  Shape<10> temp_shape;
+  for (size_t i = 0; i < in_shape.ndim(); ++i) {
+    temp_shape[i] = in_shape[i];
+  }
+
+  MSHADOW_TYPE_SWITCH(out_data.type_flag_, DType, {
+    mxnet_op::Kernel<mshadow_op::shape_kernel, gpu>::Launch(
```
Comment (reminisce) [inline src/operator/tensor/elemwise_unary_op_basic.cu]: Same as above, it's simply copying a tiny amount of data from a cpu array to a gpu array. Launching kernel is expensive and a waste of resources. You can just call `cudaMemcpyAsync` to alleviate the workload.

Comment (piiswrong) [reply src/operator/tensor/elemwise_unary_op_basic.cu]: I think this should be fine. it's unlikely to be a bottleneck

File: src/operator/tensor/elemwise_unary_op_basic.cu [hunk 20]
```diff
@@ -77,6 +77,56 @@ NNVM_REGISTER_OP(_identity_with_attr_like_rhs)
...
+                       const std::vector<TBlob>& outputs) {
+  using namespace mshadow;
+  CHECK_EQ(inputs.size(), 1U);
+  CHECK_EQ(outputs.size(), 1U);
+  CHECK_EQ(req.size(), 1U);
+  const TBlob& in_data = inputs[0];
+  const TBlob& out_data = outputs[0];
+  mshadow::Stream<gpu> *s = ctx.get_stream<gpu>();
+  const TShape& in_shape = in_data.shape_;
+  Shape<10> temp_shape;
```
Comment (piiswrong) [inline src/operator/tensor/elemwise_unary_op_basic.cu]: What if ndim is greater than 10?

Actually I agree with reminisce, you should use cudamemcopy here so that you don't need this magic number

Comment (anirudhacharya) [reply src/operator/tensor/elemwise_unary_op_basic.cu]: okay, i will fix it.

**Comment by anirudhacharya** [comment]
@reminisce @piiswrong ping for review.

File: src/operator/mshadow_op.h [hunk 21]
```diff
@@ -92,7 +92,8 @@ MXNET_UNARY_MATH_OP(identity_grad, 1);
 struct identity_with_cast {
   template<typename DTypeIn, typename DTypeOut>
   MSHADOW_XINLINE static void Map(int i, DTypeOut *out, DTypeIn *in) {
-    out[i] = DTypeOut(in[i]);
+    DTypeIn in_data = in[i];
```
Comment (reminisce) [inline src/operator/mshadow_op.h]: What's the purpose of this change?

File: src/operator/tensor/elemwise_unary_op.h [hunk 22]
```diff
@@ -388,6 +388,20 @@ void CastCompute(const nnvm::NodeAttrs& attrs,
   });
 }
 
+template<typename xpu>
+void ShapeCompute(const nnvm::NodeAttrs& attrs,
```
Comment (reminisce) [inline src/operator/tensor/elemwise_unary_op.h]: You don't need to make templates for shape and size functions based on the type of device. CPU and GPU FCompute functions are defined respectively in .cc and .cu and don't share anything.

**Comment by anirudhacharya** [comment]
@piiswrong can this be merged?

File: src/operator/tensor/elemwise_unary_op_basic.cc [hunk 23]
```diff
@@ -398,6 +398,98 @@ NNVM_REGISTER_OP(reshape_like)
 .add_argument("lhs", "NDArray-or-Symbol", "First input.")
 .add_argument("rhs", "NDArray-or-Symbol", "Second input.");
 
+void ShapeComputeCPU(const nnvm::NodeAttrs& attrs,
+                       const OpContext& ctx,
```
Comment (haojin2) [inline src/operator/tensor/elemwise_unary_op_basic.cc]: nit: alignments of all ComputeXPU functions
